File: 2021/day02.hs
module Main where
import Data.Foldable (foldl')

data Move = F Int | D Int
  deriving (Eq, Show)

parse :: String -> Maybe Move
parse ('f':xs) = Just . F $ read (drop 7 xs)
parse ('d':xs) = Just . D $ read (drop 4 xs)
parse ('u':xs) = Just . D $ -read (drop 2 xs)
parse _ = Nothing

inp :: String -> IO (Maybe [Move])
inp path = do
  c <- readFile path
  return $ mapM parse (lines c)

part1 :: Maybe [Move] -> Int
part1 Nothing = 0
part1 (Just xs) = h * d
  where (h, d) = foldl' (\(h, d) x ->
                           case x of
                             F n -> (h + n, d)
                             D n -> (h, d + n))
                 (0, 0) xs

part2 :: Maybe [Move] -> Int
part2 Nothing = 0
part2 (Just xs) = h * d
  where (h, d, _) = foldl' (\(h, d, a) x ->
                              case x of
                                F n -> (h + n, d + (a * n), a)
                                D n -> (h, d, a + n))
                    (0, 0, 0) xs

main :: IO ()
main = do
  instructions <- inp "../inputs/2021/day02.input"
  print $ part1 instructions
  print $ part2 instructions

File: 2021/day02.py
import time

start = time.time()
F = object()
D = object()


with open("../inputs/2021/day02.input", "r") as f:
    moves = []
    for line in f.read().strip().split("\n"):
        if line[0] == "f":
            moves.append((F, int(line[8:])))
        elif line[0] == "d":
            moves.append((D, int(line[5:])))
        elif line[0] == "u":
            moves.append((D, -int(line[3:])))
        else:
            raise ValueError


def part1(moves):
    h, d = 0, 0
    for move, n in moves:
        if move is F:
            h += n
        else:
            d += n
    return h * d


def part2(moves):
    h, d, a = 0, 0, 0
    for move, n in moves:
        if move is F:
            h += n
            d += a * n
        else:
            a += n
    return h * d


print(
    f"Day 02     {part1(moves):<14} {part2(moves):<14} {(time.time() - start)*1e3:>11.2f}"
)
